code/game/objects/items/devices/dogtag_detector.dm

```diff
@@ -0,0 +1,52 @@
+/obj/item/device/motiondetector/dogtag
+	name = "dogtag detector"
+	desc = "A device that detects fallen marine's dogtags."
+	icon_state = "dogtagdetector"
+	item_state = "dogtag_detector"
+	blip_type = "dogtag"
+	w_class = SIZE_SMALL
+
+/obj/item/device/motiondetector/dogtag/get_help_text()
+	. = "Red indicators on your HUD will show the location of dogtags detected by the scanner. Has two modes: slow long-range [SPAN_HELPFUL("(14 tiles)")] and fast short-range [SPAN_HELPFUL("(7 tiles)")]."
+
+/obj/item/device/motiondetector/dogtag/update_icon()
+	if (active)
+		icon_state = "[initial(icon_state)]_on_[detector_mode]"
+	else
+		icon_state = "[initial(icon_state)]"
+
+/obj/item/device/motiondetector/dogtag/scan()
+	set waitfor = 0
+	if(scanning)
+		return
+	scanning = TRUE
+	var/mob/living/carbon/human/human_user
+	if(ishuman(loc))
+		human_user = loc
+
+	var/detected_sound = FALSE
+
+	for(var/mob/living/carbon/human/sourcemob in orange(detector_range, loc))
+		var/detected
+		if(loc == null || sourcemob == null)
```

Comment:
if loc is null it's gonna be null for everyone. bail out early, not in the loop
you can also store the result of orange() and -= loc from it
then you can check z directly with `if(loc.z != sourcemob?.z) continue` and get rid of the rest